### ISSUE TITLE
Untracked files .dvc/state and .dvc/state.lock?

### ISSUE THREAD

**Issue by andrethrill** [iterative/dvc#859]
```
$ dvc --version
dvc 0.9.7+2542ce
```

After installing dvc from source, I've ran:

```
$ dvc init
$ dvc config cache.dir /path/to/cache.   # I want to use shared cache
$ dvc add [some data source raw data files]

```
_I can see the data files in the cache dir I've defined._

## Question 1

What I find it odd is that  now when I run `git status`, in the Untracked Files section I can see the files:

- `.dvc/state`
- `.dvc/state.lock`

Shouldn't this be handled by dvc automatically in `.gitignore`?

## Question 2

Ignoring the untracked files above, I've commited and pushed everything.

Now I've cloned the same repo to another directory. When I run `dvc checkout` to get the data files back from the __shared cache__ it complains that cannot find and cache file with a message like:

```
Corrupted cache file ../../dvc_caches/cache1/33/825537c2ad36adec5b48206303f3d6
Cache '33825537c2ad36adec5b48206303f3d6' not found. File 'data/raw/myfile' won't be created.
```

And now if I go to the defined cache directory, all the directories are still there but all the files disappeared:

```
$ tree .
.
├── 06
├── 09
├── 33
├── 64
├── 7d
├── 9b
├── a1
├── aa
├── be
└── e2
```

Is this a bug (maybe related with question 1)? Am I doing something wrong?

**Comment by efiop** [comment]
Hi @andrethrill !

Thank you for your feedback! Yes, indeed state and state.lock should've been added to gitignore. Fixed in #860. As to the second question, it is definitely a bug, but I suspect that it is related to the symlink link type. Could you please confirm or deny that the cache dir and the repo itself are located on different filesystems/drives?

Thank you,
Ruslan

**Comment by andrethrill** [comment]
Hi @efiop Thanks for the quick answer!

The repo and the cache are located in the same filesystems/drives.

This is how the setup looks like:

- _/datadrive/andre/repo_ # Where I originally cloned the repo, ran `dvc init` and `dvc config cache.dir /datadrive/dvc-caches/repo`
- _/datadrive/dvc-caches/repo_ # the cache itself
- _/datadrive/andre2/repo_ # where I'm cloning the same rep, running `dvc checkout` and the unexpected occurs



I'm wondering:

1 -  does the `dvc config cache.dir /datadrive/dvc-caches/repo` needs to have the path in a relative fashion rather than absolute?
2 - It was me who created the cache folder using `mkdir /datadrive/dvc-caches/repo` before running the the `dvc config` above. Maybe I need to let dvc create the folder instead of me?


I'm going to try 2. in a few moments.

**Comment by efiop** [comment]
> 1 - does the dvc config cache.dir /datadrive/dvc-caches/repo needs to have the path in a relative fashion rather than absolute?

We actually currently require an absolute path rather than relative.

> 2 - It was me who created the cache folder using mkdir /datadrive/dvc-caches/repo before running the the dvc config above. Maybe I need to let dvc create the folder instead of me?

No, it should not matter who created it.

Thank you for the info! I'm looking into it right now.

**Comment by andrethrill** [comment]
I've just tried 2. and it did not make a difference.

Does it matter where you run `dvc checkout` from? (`[repo_root]` vs `[repo_root]/data` for example)

> We actually currently require an absolute path rather than relative.

Just checking: This is, then, a change you made sometime this week right? Because this Monday(or Tuesday) I was doing a simple test and setting the cache dir as a relative path and it worked.


Thank you for your support.

**Comment by efiop** [comment]
> Does it matter where you run dvc checkout from? ([repo_root] vs [repo_root]/data for example)

Nope, it doesn't matter.

> Just checking: This is, then, a change you made sometime this week right?

Just checked and there is actually no errors/warning printed about that, so looks like we need to add one just in case.

Could you please run this script and show me the output?
```
#!/bin/bash                                                
                                                           
set -e                                                     
set -x                                                     
                                                           
rm -rf cache                                               
rm -rf myrepo1                                             
rm -rf myrepo2                                             
                                                           
mkdir cache                                                
mkdir myrepo1                                              
                                                           
pushd myrepo1                                              
git init                                                   
dvc --version                                              
dvc init -v                                                
git commit -m "Init"                                       
dvc config cache.dir $(realpath ../cache)                  
echo foo > foo                                             
dvc add foo -v                                             
ls -la                                                     
ls -i foo                                                  
stat -c "%Y" foo                                           
ls -i ../cache/d3/b07384d113edec49eaa6238ad5ff00           
stat -c "%Y" ../cache/d3/b07384d113edec49eaa6238ad5ff00    
cat .dvc/state                                             
git add .                                                  
git commit -m 'foo'                                        
popd                                                       
                                                           
git clone myrepo1 myrepo2                                  
                                                           
pushd myrepo2                                              
dvc checkout -v                                            
ls -la                                                     
ls -i foo                                                  
stat -c "%Y" foo                                           
ls -i ../cache/d3/b07384d113edec49eaa6238ad5ff00           
stat -c "%Y" ../cache/d3/b07384d113edec49eaa6238ad5ff00    
cat .dvc/state                                             
popd                                                       
```

Thanks,
Ruslan

**Comment by andrethrill** [comment]
Here's the output:

```

+ rm -rf cache
+ rm -rf myrepo1
+ rm -rf myrepo2
+ mkdir cache
+ mkdir myrepo1
+ pushd myrepo1
/datadrive/dvc_tests/myrepo1 /datadrive/dvc_tests
+ git init
Initialized empty Git repository in /datadrive/dvc_tests/myrepo1/.git/
+ dvc --version
dvc 0.9.7+08b327
+ dvc init -v
+ git commit -m Init
[master (root-commit) f259085] Init
 Committer: Ubuntu .......
Your name and email address were configured automatically based
on your username and hostname. Please check that they are accurate.
You can suppress this message by setting them explicitly. Run the
following command and follow the instructions in your editor to edit
your configuration file:

    git config --global --edit

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

 2 files changed, 8 insertions(+)
 create mode 100644 .dvc/.gitignore
 create mode 100644 .dvc/config
++ realpath ../cache
+ dvc config cache.dir /datadrive/dvc_tests/cache
+ echo foo
+ dvc add foo -v
updater is not old enough to check for updates
Cache type 'reflink' is not supported
+ ls -la
total 28
drwxrwxr-x 4 leopard leopard 4096 Jul  5 15:56 .
drwxrwxr-x 4 leopard leopard 4096 Jul  5 15:56 ..
drwxrwxr-x 3 leopard leopard 4096 Jul  5 15:56 .dvc
-rw-rw-r-- 2 leopard leopard    4 Jul  5 15:56 foo
-rw-rw-r-- 1 leopard leopard  126 Jul  5 15:56 foo.dvc
drwxrwxr-x 8 leopard leopard 4096 Jul  5 15:56 .git
-rw-rw-r-- 1 leopard leopard    3 Jul  5 15:56 .gitignore
+ ls -i foo
1572915 foo
+ stat -c %Y foo
1530806200
+ ls -i ../cache/d3/b07384d113edec49eaa6238ad5ff00
1572915 ../cache/d3/b07384d113edec49eaa6238ad5ff00
+ stat -c %Y ../cache/d3/b07384d113edec49eaa6238ad5ff00
1530806200
+ cat .dvc/state
{"1572915": {"mtime": "1530806200247715584", "md5": "d3b07384d113edec49eaa6238ad5ff00"}}+ git add .
+ git commit -m foo
[master 879350d] foo
 Committer: Ubuntu ..........
Your name and email address were configured automatically based
on your username and hostname. Please check that they are accurate.
You can suppress this message by setting them explicitly. Run the
following command and follow the instructions in your editor to edit
your configuration file:

    git config --global --edit

After doing this, you may fix the identity used for this commit with:

    git commit --amend --reset-author

 3 files changed, 9 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 foo.dvc
+ popd
/datadrive/dvc_tests
+ git clone myrepo1 myrepo2
Cloning into 'myrepo2'...
done.
+ pushd myrepo2
/datadrive/dvc_tests/myrepo2 /datadrive/dvc_tests
+ dvc checkout -v
Checking out 'foo' with cache 'd3b07384d113edec49eaa6238ad5ff00'
Cache type 'reflink' is not supported
+ ls -la
total 28
drwxrwxr-x 4 leopard leopard 4096 Jul  5 15:56 .
drwxrwxr-x 5 leopard leopard 4096 Jul  5 15:56 ..
drwxrwxr-x 2 leopard leopard 4096 Jul  5 15:56 .dvc
-rw-rw-r-- 3 leopard leopard    4 Jul  5 15:56 foo
-rw-rw-r-- 1 leopard leopard  126 Jul  5 15:56 foo.dvc
drwxrwxr-x 8 leopard leopard 4096 Jul  5 15:56 .git
-rw-rw-r-- 1 leopard leopard    3 Jul  5 15:56 .gitignore
+ ls -i foo
1572915 foo
+ stat -c %Y foo
1530806200
+ ls -i ../cache/d3/b07384d113edec49eaa6238ad5ff00
1572915 ../cache/d3/b07384d113edec49eaa6238ad5ff00
+ stat -c %Y ../cache/d3/b07384d113edec49eaa6238ad5ff00
1530806200
+ cat .dvc/state
{"1572915": {"md5": "d3b07384d113edec49eaa6238ad5ff00", "mtime": "1530806200247715584"}}+ popd
/datadrive/dvc_tests


```



I've noticed the message `Cache type 'reflink' is not supported`

**Comment by efiop** [comment]
The "Cache type 'reflink' is not supported" message is normal, your FS doesn't support it, so dvc had to fall back to hardlinks.

In the script you've run I tried to reproduce the original issue and I don't see dvc malfunctioning at all. Have you modified your data file(i.e. appended to the file inplace or something like that)? If so, that was a dvc detecting a corrupted cache file, removing it and not having the cache to checkout your new repo with.

Could you try reproducing your issue once again please? Ideally with an updated dvc from current master branch.

Thanks,
Ruslan

**Comment by andrethrill** [comment]
OK I'm still having the same problem. But I found out something interesting:

- if I use the real data files that I want to use I get them by running `cp /[original_path]/file.csv [repo_path]/data` --> This generates the odd behaviour I'm describing above
- if I create a dummy file using something like `echo mytext > [repo_path]/data` everything works flawlessly.


This starts to go beyond my scope... 

1. In terms of file systems, could something be happening under the hood when I run `cp` that makes the new copied file special (somehow) when compared to writing a new file with `echo`?
2. Are there any particular requirements for the data files to make sure MD5's are coherent? For instance, my original data files (CSV) have special characters such as é, è or à.  As far as I know this shouldn't matter, but now I'm starting to question everything...


Thank you once again for you great support @efiop .

**Comment by efiop** [comment]
> In terms of file systems, could something be happening under the hood when I run cp that makes the new copied file special (somehow) when compared to writing a new file with echo?

Yes, if you copy _over_ the existing file that is tracked by dvc. You see, in your case dvc uses hardlinks to link your data to cache, so they both share the same inode and so if you modify your data without removing it first(so that inode is changed) your cache is going to be corrupted. This is an optimization for big files that we have and this is one of the cons. We support four link types right now(reflink, hardlink, symlink, copy) and in an ideal world reflink is what we would always use, but it is not supported by all the filesystems as of today. We will probably introduce read-only-hardlinks later to provide an additional level of protection against accidental corruption. Could you please run `dvc config cache.type copy` after you run `dvc config cache.dir /path/to/cache/dir` and try going through your scenario once again so that we can make sure that you didn't corrupt your cache accidentally?

> Are there any particular requirements for the data files to make sure MD5's are coherent? For instance, my original data files (CSV) have special characters such as é, è or à. As far as I know this shouldn't matter, but now I'm starting to question everything...

This is a great point. We read data files in binary mode when computing md5 and we also do CRLF -> LF conversions to make sure that md5s are the same on unix and windows. Could you please try running these commands on your data and show us the output:
```
dos2unix mydata mydata.unix
md5sum mydata.unix
python -c 'from dvc.utils import file_md5; print(file_md5("mydata.unix")[0])'
```

Thanks,
Ruslan

**Comment by andrethrill** [comment]
@efiop sorry I was not completely clear. I have my original data files in a path where they are never modified, let's call it `orig_files_path`. Let's call the repository path `repo_path`:

- if I use the real data files that I want to use, I get them into the repositoyr by running `cp /[orig_files_path]/file.csv [repo_path]/data`. And then run `dvc add ./data/file.csv`. _After this I never modify any file, not the file in the`orig_files_path`, not the file in `[repo_path]/data` or anything in the cache_. I simply push my changes, pull them in a second path, run `dvc checkout` the odd behaviour described above occurs.
- if I create a dummy file using something like `echo mytext > [repo_path]/data/file.csv`,  then run `dvc add ./data/file.csv`, push my changes, pull them in a second path and run `dvc checkout` everything works flawlessly.

So, I believe, what your new script is doing is not what is actually happening.


> Could you please run dvc config cache.type copy after you run dvc config cache.dir /path/to/cache/dir and try going through your scenario once again so that we can make sure that you didn't corrupt your cache accidentally?

Unfortunately, nothing changed. The same error occurs.

Regarding the md5 point:

```
dos2unix mydata mydata.unix
md5sum mydata.unix
python -c 'from dvc.utils import file_md5; print(file_md5("mydata.unix")[0])'
```

Didn't you mean to calculate the md5 of the original data somewhere, for comparison? In your example you re always calculating the md5 of the .unix file.

Anyway, here's the output:

```
$ dos2unix -n production.csv production.csv.unix

$ md5sum production.csv.unix
bd9193ab817de999b526f01845609ad0  production.csv.unix

$ python -c 'from dvc.utils import file_md5; print(file_md5("production.csv.unix")[0])'
bd9193ab817de999b526f01845609ad0

$ md5sum production.csv
bd9193ab817de999b526f01845609ad0  production.csv
```

The cache never changed.

I'm really scratching my head over this.... there is something special about: either the data files, or by putting them in the repo by the `cp` command....


Thanks,
Andre

**Comment by efiop** [comment]
Ok, nevermind. I was able to reproduce. Here is a script just for the record:

```
#!/bin/bash                                     
                                                
set -e                                          
set -x                                          
                                                
rm -rf cache                                    
rm -rf myrepo1                                  
rm -rf myrepo2                                  
                                                
# NOTE: Put path to your original data file here
DATA=$(realpath file.csv)                       
                                                
echo foo > file.csv                             
unix2dos file.csv                               
                                                
mkdir cache                                     
mkdir myrepo1                                   
                                                
pushd myrepo1                                   
git init                                        
dvc --version                                   
dvc init -v                                     
git commit -m "Init"                            
dvc config cache.dir $(realpath ../cache)       
cp $DATA file.csv                               
dvc add file.csv -v                             
ls -la                                          
ls -i file.csv                                  
stat -c "%Y" file.csv                           
cat .dvc/state                                  
git add .                                       
git commit -m 'file'                            
                                                
popd                                            
                                                
git clone myrepo1 myrepo2                       
                                                
pushd myrepo2                                   
dvc checkout -v                                 
ls -la                                          
ls -i file.csv                                  
stat -c "%Y" foo.csv                            
cat .dvc/state                                  
popd                                            
```

The reason for such behavior is that the file itself is not being converted to unix format when moved to the cache. I'm already working on the fix and it is going to be released with 0.10.0 today.

Thank you so much for the feedback and your patience :)

Thanks,
Ruslan

**Comment by efiop** [comment]
@andrethrill I merged the fix. Could you please try out updated dvc and share if it worked for you or not? Feel free to reopen this issue if the problem is not solved.

Thank you!

**Comment by andrethrill** [comment]
@efiop it's with a smile on my face that I tell you I've just tested this and it seems to be working now!

I'm glad we were able to reach some conclusion and solution after digging this deep :)


I just have one last doubt that I would like to check with you: it was not clear for me how exactly `dos2unix` is used and what gets stored in the cache. Are the data files that are checked out from cache exactly equal, every single bit, to the original files I `dvc add` initially? Or are they the output of `dos2unix`?

**Comment by efiop** [comment]
@andrethrill Glad it worked for you! :) Btw, we've released 0.10.0, feel free to try it out :)

Yes, the data files that are checked out from the cache are exactly equal, every single bit, to the original files that you `dvc add` initially. The reason why I've asked you to put your data through dos2unix is because I wanted to compare dos2unix + md5sum to the md5 that is computed by dvc. In dvc we do CRLF -> LF conversion(which is what dos2unix does) when computing md5 for the data file to make sure that md5 is the same on windows and *nix. For example if you store some of your data/dependencies in the git and don't tell it to not convert CRLF <-> LF explicitly(git config core.autocrlf) you would get different md5 sum for the seemingly same file on windows and *nix.

You can verify it for yourself by running `cmp [original_path]/data.csv [repo_path]/data.csv` to verify that the checked-out data is identical and also by running `cmp [original_path]/data.csv [cache_dir]/bd/9193ab817de999b526f01845609ad0`(i'm using md5 from your production.csv, but you can retrieve it for any file from respective dvc file under outs/md5)  to make sure that the cache is identical to your original file. 

Thanks,
Ruslan
